src/vs/workbench/contrib/terminal/browser/terminalTypeAheadAddon.ts

```diff
@@ -1381,16 +1391,22 @@ export class TypeAheadAddon extends Disposable implements ITerminalAddon {
 		// the user gave input, and mark all additions before that as tentative.
 		const actualY = buffer.baseY + buffer.cursorY;
 		if (actualY !== this.lastRow?.y) {
-			this.lastRow = { y: actualY, startingX: buffer.cursorX, charState: CharPredictState.Unknown };
+			this.lastRow = { y: actualY, startingX: buffer.cursorX, endingX: buffer.cursorX, charState: CharPredictState.Unknown };
 		} else {
 			this.lastRow.startingX = Math.min(this.lastRow.startingX, buffer.cursorX);
+			this.lastRow.endingX = Math.max(this.lastRow.endingX, buffer.cursorX);
```

Comment:
This can actually be the cursor x, which will allow left/right navigation in predicted characters

```suggestion
			this.lastRow.endingX = Math.max(this.lastRow.endingX, this.timeline.getCursor(buffer).x);
```

We don't need to do this for startingX, we will never apply a prediction (backspace or left arrow) that goes beyond the startingX, but for endingX the character predictions should increase the navigable area.

Reply:
In better words -- breaking the startingX boundary will always require a round trip to the pty, which will be relatively slow. But the endingX boundary will happen on key type, so we should use the internal cursor position to avoid incorrectly adding tentative boundaries if there are writes which are not yet flushed.

Technically the startingX can use the internal cursor position as well, but that will be about equivalent.

Reply:
That change still gives me the behavior of the cursor going beyond where it is allowed.

Reply:
Clarification: 

> That change still gives me the behavior of the cursor going beyond where it is allowed

This only happens when I _hold down right arrow_. If I just hit it once in a fresh terminal, everything works as expected.

Reply:
Oh, I think there's an off-by-one in `addRightNavigating`, it should be `>= this.lastRow!.endingX - 1`, since we're seeing it we can move _ahead_ to that position at which we've never been before.